File: Python/tcsproblemE.py
from collections import defaultdict

def calculate_interest(principal, weeks):
    interest_rate = 0.12 / 52  
    interest = principal * (1 + interest_rate)**weeks - principal
    return round(interest, 2)

def settle_expenses(expenses, balances):
    total_expense = sum(expenses.values())
    share_per_person = total_expense / len(balances)

    for person, amount in balances.items():
        balances[person] -= share_per_person

def reconcile_loans(loans, balances):
    sorted_loans = sorted(loans, key=lambda x: (loans[x]['lender'], x))

    for borrower in sorted_loans:
        for lender, amount in loans[borrower]['amounts'].items():
            if balances[borrower] >= amount:
                balances[borrower] -= amount
                balances[lender] += amount
                loans[borrower]['amounts'][lender] = 0
            else:
                loans[borrower]['amounts'][lender] -= balances[borrower]
                balances[lender] += balances[borrower]
                balances[borrower] = 0
                break

def process_transaction(transaction, expenses, loans, balances, week):
    try:
        if transaction[0] == 'L':
            lender, borrower, amount = transaction[2:].split('/')
            amount = int(amount)
            
            if borrower not in loans:
                loans[borrower] = {'lender': lender, 'amounts': defaultdict(int)}
            
            loans[borrower]['amounts'][lender] += amount
            loans[borrower]['week'] = week
        else:
            payer, amounts = transaction.split('/', 1)
            amounts = amounts.split('/')
            paid_amount = int(amounts[0])
            recipients = amounts[1:]

            for recipient in recipients:
                expenses[payer] += paid_amount / len(recipients)
                balances[recipient] += paid_amount / len(recipients)

                if payer in loans and loans[payer]['week'] == week:
                    for lender, amount in loans[payer]['amounts'].items():
                        if balances[payer] >= amount:
                            balances[payer] -= amount
                            balances[lender] += amount
                            loans[payer]['amounts'][lender] = 0
                        else:
                            loans[payer]['amounts'][lender] -= balances[payer]
                            balances[lender] += balances[payer]
                            balances[payer] = 0
                            break
    except (ValueError, IndexError):
        raise ValueError("Invalid transaction format: " + transaction)

def main():
    try:
        n = int(input())
        transactions = [input().strip() for _ in range(n)]

        week = 1
        expenses = defaultdict(int)
        loans = {}
        balances = defaultdict(int)

        for transaction in transactions:
            process_transaction(transaction, expenses, loans, balances, week)
            week += 1

            for borrower in loans.copy():
                if week - loans[borrower]['week'] > 1:
                    for lender, amount in loans[borrower]['amounts'].items():
                        interest = calculate_interest(amount, week - loans[borrower]['week'])
                        loans[borrower]['amounts'][lender] += interest
                        loans[borrower]['week'] = week

        settle_expenses(expenses, balances)
        reconcile_loans(loans, balances)

        dues = [(payer, recipient, round(amount, 2)) for payer, recipient, amount in sorted([(payer, recipient, amount) for payer, recipient, amount in balances.items() if amount != 0])]

        if dues:
            for payer, recipient, amount in dues:
                print(f"{payer}/{recipient}/{amount}")
        else:
            print("NO DUES.")

    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: c/c++/tcsproblemE.cpp
#include <iostream>
#include <iomanip>
#include <sstream>
#include <unordered_map>
#include <map>
#include <vector>
#include <cmath>
#include <algorithm>

class Loan {
public:
    std::string lender;
    std::unordered_map<std::string, int> amounts{};
    int week;
};

double calculateInterest(int principal, int weeks) {
    double interestRate = 0.12 / 52;
    double interest = principal * (std::pow((1 + interestRate), weeks) - 1);
    return std::round(interest * 100) / 100;
}

void settleExpenses(std::unordered_map<std::string, int>& expenses, std::unordered_map<std::string, double>& balances) {
    double totalExpense = 0;
    for (const auto& person : balances) {
        totalExpense += person.second;
    }

    double sharePerPerson = totalExpense / balances.size();

    for (auto& person : balances) {
        person.second -= sharePerPerson;
    }
}

void reconcileLoans(std::vector<std::tuple<std::string, std::string, double>>& dues,
                    std::unordered_map<std::string, Loan>& loans,
                    std::unordered_map<std::string, double>& balances) {
    for (const auto& due : dues) {
        std::string borrower = std::get<0>(due);
        std::string lender = std::get<1>(due);
        double amount = std::get<2>(due);

        if (loans.find(borrower) != loans.end()) {
            for (const auto& loanAmount : loans[borrower].amounts) {
                if (balances[borrower] >= loanAmount.second) {
                    balances[borrower] -= loanAmount.second;
                    balances[loanAmount.first] += loanAmount.second;
                    loans[borrower].amounts[loanAmount.first] = 0;
                } else {
                    loans[borrower].amounts[loanAmount.first] -= balances[borrower];
                    balances[loanAmount.first] += balances[borrower];
                    balances[borrower] = 0;
                    break;
                }
            }
        } else if (!lender.empty()) {
            balances[borrower] -= amount;
            balances[lender] += amount;
        }
    }
}

void processTransaction(const std::string& transaction, std::unordered_map<std::string, int>& expenses,
                        std::vector<std::tuple<std::string, std::string, double>>& dues,
                        std::unordered_map<std::string, Loan>& loans, std::unordered_map<std::string, double>& balances, int week) {
    try {
        if (transaction[0] == 'L') {
            std::istringstream iss(transaction.substr(2));
            std::string lender, borrower;
            int amount;
            char delimiter;
            iss >> lender >> delimiter >> borrower >> delimiter >> amount;

            if (loans.find(borrower) == loans.end()) {
                loans[borrower] = Loan{.lender = lender, .week = week};
            }

            loans[borrower].amounts[lender] += amount;
        } else {
            std::istringstream iss(transaction.substr(2));
            std::string payer, recipients;
            int paidAmount;
            char delimiter;
            iss >> payer >> delimiter >> paidAmount >> recipients;

            std::istringstream recipientsStream(recipients);
            std::string recipient;
            while (getline(recipientsStream, recipient, '/')) {
                expenses[payer] += paidAmount / recipients.size();
                balances[recipient] += paidAmount / recipients.size();

                if (loans.find(payer) != loans.end() && loans[payer].week == week) {
                    for (const auto& amount : loans[payer].amounts) {
                        if (balances[payer] >= amount.second) {
                            balances[payer] -= amount.second;
                            balances[amount.first] += amount.second;
                            loans[payer].amounts[amount.first] = 0;
                        } else {
                            loans[payer].amounts[amount.first] -= balances[payer];
                            balances[amount.first] += balances[payer];
                            balances[payer] = 0;
                            break;
                        }
                    }
                }
            }
        }
    } catch (const std::exception& e) {
        throw std::invalid_argument("Invalid transaction format: " + transaction);
    }
}

int main() {
    try {
        int n;
        std::cin >> n;
        std::cin.ignore();

        std::vector<std::string> transactions;
        for (int i = 0; i < n; ++i) {
            std::string transaction;
            std::getline(std::cin, transaction);
            transactions.push_back(transaction);
        }

        int week = 1;
        std::unordered_map<std::string, int> expenses;
        std::unordered_map<std::string, Loan> loans;
        std::unordered_map<std::string, double> balances;
        std::vector<std::tuple<std::string, std::string, double>> dues;

        for (const auto& transaction : transactions) {
            processTransaction(transaction, expenses, dues, loans, balances, week);
            ++week;
        }

        settleExpenses(expenses, balances);
        reconcileLoans(dues, loans, balances);

        std::sort(dues.begin(), dues.end(), [](const auto& a, const auto& b) {
            return std::get<0>(a) < std::get<0>(b);
        });

        bool hasDues = false;
        for (const auto& due : dues) {
            if (std::get<2>(due) != 0) {
                std::cout << std::get<0>(due) << "/" << std::get<1>(due) << "/" << std::get<2>(due) << std::endl;
                hasDues = true;
            }
        }

        if (!hasDues) {
            std::cout << "NO DUES." << std::endl;
        }

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}
